### PR TITLE
New template for Spanner changestreams to Elasticsearch

### PR DESCRIPTION
**Pull request by ggprod**
Resolves issue #413

### CODE REVIEW

File: v2/elasticsearch-common/src/main/java/com/google/cloud/teleport/v2/elasticsearch/options/ElasticsearchWriteOptions.java [hunk 1]
```diff
@@ -73,4 +73,49 @@ public interface ElasticsearchWriteOptions extends PipelineOptions {
   Long getMaxRetryDuration();
 
   void setMaxRetryDuration(Long maxRetryDuration);
+
+  @Description("Gcs path to javascript udf source for index extraction function")
```
Comment (an2x) [inline v2/elasticsearch-common/src/main/java/com/google/cloud/teleport/v2/elasticsearch/options/ElasticsearchWriteOptions.java:77]: Please make the changes requested in PR #399 here too (e.g. change parameter descriptions) and let me know when done, I'll do another pass.

**Review by an2x — CHANGES_REQUESTED** [review]

File: v2/elasticsearch-common/src/main/java/com/google/cloud/teleport/v2/elasticsearch/options/ElasticsearchWriteOptions.java (hunk 1 above)
Comment (ggprod) [reply v2/elasticsearch-common/src/main/java/com/google/cloud/teleport/v2/elasticsearch/options/ElasticsearchWriteOptions.java]: @an2x ok, I've rebased this PR against the lastest main (that has the merged changes from the PR #399

File: v2/googlecloud-to-elasticsearch/src/test/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearchTest.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+          "{\"TableName\":\"testTable\",\"AccountId\":\"AccountId2\",\"UserId\":\"UserId1\",\"IsDelete\":true}",
+          "{\"TableName\":\"testTable\",\"AccountId\":\"AccountId1\",\"UserId\":\"UserId2\",\"IsDelete\":true}",
+          "{\"TableName\":\"testTable\",\"AccountId\":\"AccountId2\",\"UserId\":\"UserId2\",\"IsDelete\":true}");
+
+  @Rule public final transient TestPipeline pipeline = TestPipeline.create();
+  @Rule public ExpectedException exceptionRule = ExpectedException.none();
+
+  /** Test the {@link SpannerChangeStreamsToElasticsearch} pipeline end-to-end. */
+  @Test
+  public void testSpannerChangeStreamsToElasticsearchE2E() {
```
Comment (an2x) [inline v2/googlecloud-to-elasticsearch/src/test/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearchTest.java:154]: This is not really and end-to-end test as it only tests `DataChangeRecordToJsonFn`.
Please at the minimum rename this test case, or make `DataChangeRecordToJsonFn` a standalone class and add a separate test class for that.
(And testing the template itself E2E is hard or impossible in a unit test. It's OK to not have any E2E tests for now. We should move integration tests to github soon. Adding an E2E integration test after that would make sense).

File: v2/googlecloud-to-elasticsearch/src/main/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearch.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+      element
+          .getMods()
+          .forEach(
+              mod -> {
+                JSONObject keysJsonExtended = new JSONObject(mod.getKeysJson());
+                keysJsonExtended.put("TableName", tableName);
+                if (modType == ModType.INSERT || modType == ModType.UPDATE) {
+                  JSONObject newValuesJson = new JSONObject(mod.getNewValuesJson());
+                  // add any properties that are present in newValuesJson to the set of keys to make
+                  // a partial (or complete) row reflecting
```
Comment (an2x) [inline v2/googlecloud-to-elasticsearch/src/main/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearch.java:159]: nit: unnecessary line break.

**Review by an2x — CHANGES_REQUESTED** [review]

File: v2/googlecloud-to-elasticsearch/src/test/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearchTest.java (hunk 2 above)
Comment (ggprod) [reply v2/googlecloud-to-elasticsearch/src/test/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearchTest.java]: I renamed the test case function.. is that ok for now or you'd prefer the second approach of a standalone class for `DataChangeRecordToJsonFn`?

File: v2/googlecloud-to-elasticsearch/src/main/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearch.java (hunk 3 above)
Comment (ggprod) [reply v2/googlecloud-to-elasticsearch/src/main/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearch.java]: I assumed you meant the comment and tried to resolve by combining.. was that what you meant?

File: v2/googlecloud-to-elasticsearch/src/test/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearchTest.java (hunk 2 above)
Comment (an2x) [reply v2/googlecloud-to-elasticsearch/src/test/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearchTest.java]: Thanks, either works!

**Review by an2x — DISMISSED** [review]

**Comment by bvolpato** [comment]
cc @nancyxu123

File: v2/googlecloud-to-elasticsearch/src/main/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearch.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        ? options.getProject()
+        : options.getSpannerProjectId();
+  }
+
+  public static PipelineResult run(SpannerChangeStreamsToElasticsearchOptions options) {
+    options.setStreaming(true);
+    options.setEnableStreamingEngine(true);
+    options.setAutoscalingAlgorithm(
+        DataflowPipelineWorkerPoolOptions.AutoscalingAlgorithmType.NONE);
+    // Spanner change stream updates would require partial updates when writing to Elasticsearch
```
Comment (nancyxu123) [inline v2/googlecloud-to-elasticsearch/src/main/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearch.java:69]: Is this something specific to the ElasticSearch template? What effect would this have?

File: v2/googlecloud-to-elasticsearch/src/main/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearch.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+  private static String getProjectId(SpannerChangeStreamsToElasticsearchOptions options) {
+    return options.getSpannerProjectId().isEmpty()
+        ? options.getProject()
+        : options.getSpannerProjectId();
+  }
+
+  public static PipelineResult run(SpannerChangeStreamsToElasticsearchOptions options) {
+    options.setStreaming(true);
+    options.setEnableStreamingEngine(true);
+    options.setAutoscalingAlgorithm(
```
Comment (nancyxu123) [inline v2/googlecloud-to-elasticsearch/src/main/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearch.java:67]: We don't need to disable autoscaling by default anymore.

File: v2/googlecloud-to-elasticsearch/src/test/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearchTest.java [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+  private static final List<Mod> insertMods =
+      ImmutableList.of(
+          new Mod(
+              "{\"UserId\": \"UserId1\", \"AccountId\": \"AccountId1\"}",
+              null,
+              "{\"LastUpdate\": \"2022-09-27T11:30:00.123456Z\", \"Balance\": 1000}"),
+          new Mod(
+              "{\"UserId\": \"UserId1\", \"AccountId\": \"AccountId2\"}",
+              null,
+              "{\"LastUpdate\": \"2022-09-27T11:30:00.123456Z\", \"Balance\": 3000}"),
```
Comment (nancyxu123) [inline v2/googlecloud-to-elasticsearch/src/test/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearchTest.java:54]: FYI I believe you can mock calls into Spanner if you want to test the pipeline e2e

File: v2/googlecloud-to-elasticsearch/src/main/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearch.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+                        if (newValuesJson.has(col.getName())) {
+                          keysJsonExtended.put(col.getName(), newValuesJson.get(col.getName()));
+                        }
+                      });
+                } else {
+                  // For DELETE mod, need the keys and a property indicating it is a delete
+                  keysJsonExtended.put("IsDelete", true);
+                }
+
+                output.output(keysJsonExtended.toString());
```
Comment (nancyxu123) [inline v2/googlecloud-to-elasticsearch/src/main/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearch.java:171]: Is the goal to output one JSON string per each data change record mod?

Would you be able to reuse this transformation? 

https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/42594e181db9a428246b2372924f9c4510e8527a/v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/spannerchangestreamstobigquery/SpannerChangeStreamsToBigQuery.java#L367

**Pull request closed without merging**